adapter: acquire read holds before making assumptions about collection sinces

aljoscha opened the pull request:
This is preparatory work for #24845, where we
factor a `StorageCollections` out of present-day `StorageController`.
And this `StorageCollections` will be shareable and will come with the
possibility of fontiers advancing concurrently.

We therefore need to guard places where we make assumptions about
frontiers by first acquiring read holds. This will "lock" sinces in
place where they currently are and then look at those sinces.

### Tips for reviewer

**NOTE:** There's a known bug with dropping `ReadHolds` that I still have to figure out, and I have to rebase across a slightly more involved conflict. But I wanted to get the basic structure out for review now, to see how you like and and if you prefer one of the approaches I mention below.

I implemented two solutions which I'll call _static read holds_ and _runtime read holds_ (or _dynamic read holds_). The former is everything up to the last commit, and the last commit changes it to the dynamic approach. So if you just look at this PR here on GH, you will only see the dynamic approach.

This is the commit message from that last commit:

```
adapter: dynamic lifetimes for ReadHolds 

As opposed to the previous approach, which I'd retroactively call static
lifetimes for read holds.

Before, we were using ReadHoldsGuard to ensure that ReadHolds were being
dropped when the guard went out of scope. This made for a clunky API,
though.

Now, we give each ReadHolds a channel on which it sends itself when it
is dropped. And the Coordinator works off that channel and releases
dropped read holds.

The new approach will make tracing/debugging problems slightly harder,
because there is more messaging and more state going around. But I think
we will need something like this for a more concurrent, future
Coordinator where we can acquire read holds, ferry them off to other
tasks and eventually release them there, without mutable access the the
Coordinator.
```

@teskje Added you as a reviewer because it's interesting pre-requisites for the controller refactoring PR.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

jkosh44 commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -1098,14 +1108,39 @@ impl Coordinator {
+            if let Some(read_holds) = read_holds {
+                self.release_read_holds(vec![read_holds]);
+            }
+            // drop(read_holds_guard);
```
I think you left this in by accident?

jkosh44 commented on src/adapter/tests/timestamp_selection.rs:
```diff
@@ -26,6 +27,8 @@ use mz_storage_types::sources::Timeline;
 use serde::{Deserialize, Serialize};
 use timely::progress::Antichain;
 
+use mz_adapter::ReadHolds;
```
```suggestion
use mz_adapter::ReadHolds;
```

jkosh44 commented on src/adapter/src/coord/timestamp_selection.rs:
```diff
@@ -597,8 +637,15 @@ impl Coordinator {
+
+                // Make sure we release our holds!
+                drop(tmp_read_holds);
+                // self.release_read_holds(vec![tmp_read_holds]);
```
```suggestion
```

jkosh44 commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -37,21 +38,24 @@ use timely::progress::Antichain;
+#[derive(Serialize)]
 pub struct ReadHolds<T> {
     pub holds: HashMap<Antichain<T>, CollectionIdBundle>,
+    // We only use Serialize for debugging.
```
Should we just replace the `Serialize` usages with `Debug`?

aljoscha replied:
I think a bunch of `Serialize` bounds were added for the "dump coordinator state" work that @ParkMyCar did. That uses `Serialize` to dump a json view of coordinator state, or something like it.

teskje replied:
I think the `Serialize` impl for `ReadHolds` is unneeded since https://github.com/MaterializeInc/materialize/pull/26237

ParkMyCar replied:
+1 you should be able to get rid of the `Serialize` bound

jkosh44 commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -148,75 +156,66 @@ impl<T: Eq + Hash + Ord> ReadHolds<T> {
-    }
-}
+        let drop_tx = owned_self.dropped_read_holds_tx.take();
+        let drop_tx = if let Some(drop_tx) = drop_tx {
+            drop_tx
+        } else {
+            panic!("trying to drop ReadHolds with holds on {:?} without a sender for dropped read holds", owned_self.holds.values());
+        };
```
Totally subjective, but I think you can replace this with something like 
```Rust
let drop_tx = drop_tx
    .unwrap_or_else(|| panic!("trying to drop ReadHolds with holds on {:?} without a sender for dropped read holds", owned_self.holds.values()));
```

jkosh44 commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -37,21 +38,24 @@ use timely::progress::Antichain;
     pub holds: HashMap<Antichain<T>, CollectionIdBundle>,
+    // We only use Serialize for debugging.
+    #[serde(skip_serializing)]
+    dropped_read_holds_tx: Option<tokio::sync::mpsc::UnboundedSender<ReadHolds<T>>>,
```
Is there a reason to send the whole `ReadHolds` struct instead of just sending a `HashMap<Antichain<T>, CollectionIdBundle>`? It feels a bit odd that we send a channel over the channel, but we have to remember that the sent channel is `None` (if I'm understanding that correctly).

aljoscha replied:
Yeah, sorry you had to see this. This is the bug I still have to fix. I'm sending a `ReadHolds` because that's what `release_read_holds(...)` takes. I had planned this approach for fixing it:

- Change `ReadHolds` back to what it was before.
- Add a new `DroppableReadHolds` that has a `ReadHolds` field and the channel, plus a `Deref` impl for accessing the underlying `ReadHolds`. That way I get rid of most of the cleaning-out-the-channel shenanigans.

jkosh44 commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -783,5 +785,11 @@ impl crate::coord::Coordinator {
             }
         }
+
+        for mut read_holds in read_holdses {
+            // Make sure the Drop impl does not send this along for releasing!
+            read_holds.dropped_read_holds_tx.take();
+            read_holds.holds.clear();
+        }
```
Related to another one of my comments, but if this method just had `read_holdses: Vec<HashMap<Antichain<Timestamp>, CollectionIdBundle>>` as input, then I don't think we'd need to worry about this.

aljoscha replied:
I wasn't sure about changing this signature. I'm leaving in place a bunch of explicit `release_read_holds()` calls that this would break.

I _could_ of course remove all of those and purely rely on the Drop impl. What would you prefer?

aljoscha replied:
It's changed in the PR now, because keeping the explicit `release_read_holds()` is tricky. There's now a split between `InternalReadHold` and `ReadHold`, which makes something explicit that has been there for a while in the adapter: there's `ReadHolds` that we create when creating a collection, and we keep those around forever and regularly call `update_read_hold())` (which downgrades it, naturally!), and there's `ReadHolds` that we acquire for peeks and such.

With this PR, the former is now `InternalReadHold` and the latter is just `ReadHold` (which is droppable and wraps a `InternalReadHold`.

The split also shows up in the fact that we have two different places where we create `ReadHolds`:

1. for short-lived ones: https://github.com/MaterializeInc/materialize/blob/eca245f1e55bd9fd7c4bda9297107548444563d1/src/adapter/src/coord/read_policy.rs#L553
2. for the long-lived ones: https://github.com/MaterializeInc/materialize/blob/eca245f1e55bd9fd7c4bda9297107548444563d1/src/adapter/src/coord/read_policy.rs#L294

The code path for the long-lived holds also does some initialization, but then it does basically the same thing as `acquire_read_holds`. I didn't want to mess with this too much in this PR, though.

jkosh44 commented on src/adapter/src/coord.rs:
```diff
@@ -1261,6 +1263,9 @@ pub struct Coordinator {
     /// Channel for strict serializable reads ready to commit.
     strict_serializable_reads_tx: mpsc::UnboundedSender<(ConnectionId, PendingReadTxn)>,
 
+    /// Channel for returning/releasing [ReadHolds](ReadHolds).
+    dropped_read_holds_tx: mpsc::UnboundedSender<ReadHolds<Timestamp>>,
```
Can we just re-use `internal_cmd_tx` and send the drop messages over that?

aljoscha replied:
Yep, that seems like a good idea. I had first modeled it like the other special-purpose channels and it evolved from there... 😅

aljoscha replied:
I added a comment in the code to say why I think this might be good to keep:

```
    /// Channel for returning/releasing [InternalReadHolds](InternalReadHolds).
    ///
    /// We're using a special purpose channel rather than using
    /// `internal_cmd_tx` so that we can control the priority of working off
    /// dropped read holds. If we sent them as [Message] on the internal cmd
    /// channel, these would always get top priority, which is not necessary.
```

wdyt?

jkosh44 replied:
> wdyt?

I think that's reasonable.

jkosh44 reviewed (commented):
Some of the comments are outdated on the static approach and probably don't make sense with the dynamic approach.

teskje commented on src/adapter/src/coord.rs:
```diff
@@ -2210,17 +2224,26 @@ impl Coordinator {
 
-    /// Returns an `as_of` suitable for bootstrapping the given index dataflow.
+    /// Returns an `as_of` suitable for bootstrapping the given index dataflow,
+    /// along with a [ReadHolds] that ensures the since of involved collections
```
```suggestion
    /// along with a [ReadHolds] that ensures the sinces of involved collections
```

teskje commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -20,7 +20,9 @@
 //! `mz_ore` wrapper either.
 #![allow(clippy::disallowed_types)]
 
+use core::panic;
```
Stray?

teskje commented:
If the dynamic read holds approach is the way you want to pursue (which I think you should!), how about removing the static read holds approach from this PR? It makes the commit-by-commit review quite a bit more tedious and I'm not sure the context is really needed.

aljoscha commented on src/adapter/src/coord/read_policy.rs:
```diff
@@ -272,6 +275,94 @@ impl crate::coord::Coordinator {
+                        let read_frontier = collection.read_capability().clone();
+                        ReadPolicy::NoPolicy {
+                            initial_since: read_frontier,
+                        }
```
It's complicated: sometimes it's used as a "I don't know, but keep the since at this while we figure it out" policy. But you're right, I will change it to just use `ValidFrom` here.

aljoscha commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -998,13 +999,24 @@ impl Coordinator {
+        // TODO: Maybe in the future, pass those holds on to compute, to
+        // hold on to them and downgrade when possible?
+        let read_holds_guard = self
+            .acquire_read_holds(mz_repr::Timestamp::minimum(), &id_bundle, false)
```
:+1:

aljoscha commented on src/adapter/src/coord/timestamp_selection.rs:
```diff
@@ -203,6 +204,25 @@ impl TimestampProvider for Coordinator {
+    ) -> ReadHolds<mz_repr::Timestamp> {
+        let read_holds = self
+            .acquire_read_holds(mz_repr::Timestamp::minimum(), id_bundle, false)
+            .expect("can acquire read holds");
```
I think even that `precise = true` parameter is a stop-gap solution for something else. I believe Gabor added it to make some edge cases with `REFRESH EVERY` work.

teskje replied:
Yep, that's also my understanding. So perhaps the better solution would have been to introduce a new `acquire_precise_read_holds` method that wraps an infallible `acquire_read_holds`, instead of adding a parameter  that influences the expected return type.

aljoscha commented:
> If the dynamic read holds approach is the way you want to pursue (which I think you should!), how about removing the static read holds approach from this PR? It makes the commit-by-commit review quite a bit more tedious and I'm not sure the context is really needed.

Thanks for the thorough review, I think I addressed all your comments. I also did remove all vestiges of the "static" approach, though I have the in a branch if anyone wants to compare that to the approach in this PR.

teskje commented on src/adapter/src/coord/sequencer/inner/peek.rs:
```diff
@@ -1044,52 +1044,60 @@ impl Coordinator {
+
+                    &timedomain_bundle
+                } else {
+                    tracing::debug!(session = %session.conn_id(), ?timeline_context, ?source_ids, "NOT in multi txn");
```
(There are a couple more DEBUG calls below as well, to which this also applies.)

teskje reviewed (approved):
LGTM on a high level! I'll leave the detailed review to the adapter folks.

nrainer-materialize commented:
I triggered a [nightly](https://buildkite.com/materialize/nightlies/builds/7204) and [coverage build](https://buildkite.com/materialize/coverage/builds/401) and will report the results when they are ready.

jkosh44 reviewed (approved):
LGTM

aljoscha commented:
@nrainer-materialize those coverage and nightly builds don't look good, but I can't tell if that's my changes or not

nrainer-materialize commented:
> @nrainer-materialize those coverage and nightly builds don't look good, but I can't tell if that's my changes or not

No, it's actually not that bad!

Nightly:
* product-limits fails on main as well (https://github.com/MaterializeInc/materialize/issues/26404)
* k8s-recovery is sometimes flaky, I triggered a further retry
* same for parallel-workload

Failures in the coverage build are not expressive due to different timings caused by coverage recording.

nrainer-materialize commented:
✅ Coverage results look good; no relevant test gaps!
